Implement Fetch Offer Infrastructure

Steffengreiner opened the pull request:
**Description of changes**
This PR connects the previously introduced Fetch Offer Use Case(See this [PR](https://github.com/qbicsoftware/offer-manager-2-portlet/pull/344)) to the Create Offer and OfferOverview Components.
This is necessary since both components allow the independent download of the generated Offer which should include additional information not stored in the database.

sven1103 commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/components/offer/create/CreateOfferController.groovy:
```diff
@@ -77,4 +80,13 @@ class CreateOfferController {
         }
     }
+    /**
+     * This method calls the Fetch Offer use case with the provided OfferId
```
Triggers the `FetchOffer` use case on execution.

sven1103 commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/components/offer/create/CreateOfferController.groovy:
```diff
@@ -77,4 +80,13 @@ class CreateOfferController {
+    /**
+     * This method calls the Fetch Offer use case with the provided OfferId
+     *
+     * @param offerId The OfferId of the Offer to be retrieved
```
```suggestion
     * @param offerId The identifier of the offer to be fetched
```

sven1103 commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/components/offer/create/CreateOfferPresenter.groovy:
```diff
@@ -50,4 +51,8 @@ class CreateOfferPresenter implements CreateOfferOutput {
 
+    @Override
+    void fetchedOffer(Offer fetchedOffer) {
+        this.createOfferViewModel.savedOffer = Optional.ofNullable(fetchedOffer)
```
```suggestion
        this.createOfferViewModel.savedOffer = Optional.of(fetchedOffer)
```

sven1103 replied:
Fail fast principle.

sven1103 replied:
I would also expect that the use case won't proxy a null object reference, that would be really bad.

sven1103 commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/components/offer/create/OfferOverviewView.groovy:
```diff
@@ -46,20 +46,22 @@ class OfferOverviewView extends VerticalLayout{
-                log.error(e.message)
-                log.error(e.stackTrace.join("\n"))
-            }
+                try {
+                    createOfferController.fetchOffer(offer.identifier)
+                    addOfferResource(offer)
+                } catch (Exception e) {
+                    log.error("Unable to create the offer PDF resource.")
+                    log.error(e.message)
+                    log.error(e.stackTrace.join("\n"))
+                }
```
over indentation.

Steffengreiner replied:
Nice catch 👍

sven1103 commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/components/offer/create/CreateOfferViewModel.groovy:
```diff
@@ -47,8 +48,9 @@ class CreateOfferViewModel {
-    double totalPrice = 0
+    @Bindable double totalPrice = 0
+
+    Optional<Offer> savedOffer
```
```suggestion
    Optional<Offer> savedOffer = Optional.empty()
```

sven1103 commented on offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/components/offer/create/OfferOverviewView.groovy:
```diff
@@ -202,16 +204,19 @@ class OfferOverviewView extends VerticalLayout{
-        currentFileDownloader.extend(downloadOffer)
-        downloadOffer.setEnabled(true)
+        //Check if an Offer has been saved.
+        if (createOfferViewModel.savedOffer) {
+            removeExistingResources()
+            // Then we create a new PDF resource ...
+            OfferToPDFConverter converter = new OfferToPDFConverter(createOfferViewModel.savedOffer.get())
+            StreamResource offerResource = new StreamResource((StreamResource.StreamSource res) -> {
+                return converter.getOfferAsPdf()
+            }, "${offer.identifier.toString()}.pdf")
+            // ... and attach it to the download button
+            currentFileDownloader = new FileDownloader(offerResource)
+            currentFileDownloader.extend(downloadOffer)
+            downloadOffer.setEnabled(true)
+        }
```
```suggestion
        removeExistingResources()
        if (!createOfferViewModel.savedOffer.isPresent()) {
            return
        }
        // Then we create a new PDF resource ...
        OfferToPDFConverter converter = new OfferToPDFConverter(createOfferViewModel.savedOffer.get())
        StreamResource offerResource = new StreamResource((StreamResource.StreamSource res) -> {
            return converter.getOfferAsPdf()
        }, "${offer.identifier.toString()}.pdf")
        // ... and attach it to the download button
        currentFileDownloader = new FileDownloader(offerResource)
        currentFileDownloader.extend(downloadOffer)
        downloadOffer.setEnabled(true)
        }
```

sven1103 commented on offer-manager-domain/src/main/groovy/life/qbic/business/offers/fetch/FetchOffer.groovy:
```diff
@@ -32,21 +32,21 @@ class FetchOffer implements FetchOfferInput {
         catch (DatabaseQueryException queryException) {
             log.error(queryException.message)
-            output.failNotification("Could not retrieve Offer with OfferId $offerId from the Database")
+            output.failNotification("Could not retrieve Offer with OfferId " + offerId.toString() + " from the Database")
```
```suggestion
            output.failNotification("Could not retrieve Offer with OfferId ${offer.toString()} from the Database")
```

sven1103 replied:
and the same for the others pls

sven1103 reviewed (changes requested):
Nearly there :)

sven1103 commented:
Changelog!!

Steffengreiner commented:
@sven1103 I incorporated your feedback, introduced a seperate instance of the Fetch Offer process for the Update Offer Use Case and will update the changelog in more detail in the PR, which will address the changes in the HTML document. 
Thank you for your detailed suggestions.  👍

sven1103 reviewed (commented):
looks good on the code level, i will check out and test the behaviour now
